File: RoundRobin.rb
# Algoritmo de planificación por turnos (Round Robin)
# @faustinoaq Sept, 2016

class Proceso
  attr_accessor :pcb, :id, :instru, :estado, :posic
  def initialize(pcb, id, instru, estado, posic)
    @pcb = pcb
    @id = id
    @instru = instru
    @estado = estado
    @posic = posic
  end
end

class RoundRobin
  def initialize(n, quantum, archivo)
    @n = n
    @quantum = quantum
    @archivo = archivo
    @procesos = (1..@n).map do |i|
      Proceso.new("P#{i}", "0#{i}", 100, "N", i)
    end
    titulo_buffer
    cola_listo
    procesar_cola
    puts "#{@archivo} guardado correctamente"
  end

  private def titulo_buffer
    @buffer = ["Algoritmo de planificación Round Robin\n"]
    @buffer << "N(Nuevo) L(Listo) E(Ejecución) T(Terminado)\n"
    @buffer << "Cantidad de procesos: #{@n}\n"
    @buffer << "Quantum: #{@quantum}\n"
    guardar_buffer
  end

  private def guardar_buffer
    @buffer << "\nProceso"
    @procesos.each do |p|
      @buffer << "\t#{p.pcb}"
    end
    @buffer << "\nID"
    @procesos.each do |p|
      @buffer << "\t#{p.id}"
    end
    @buffer << "\nInstruc"
    @procesos.each do |p|
      @buffer << "\t#{p.instru}"
    end
    @buffer << "\nEstado"
    @procesos.each do |p|
      @buffer << "\t#{p.estado}"
    end
    @buffer << "\nPosic"
    @procesos.each do |p|
      @buffer << "\t#{p.posic}"
    end
    @buffer << "\n"
    guardar_archivo
  end

  private def cola_listo
    @procesos.each do |p|
      p.estado = "L"
    end
    guardar_buffer
  end

  private def procesar_cola
    until procesos_terminados
      @procesos.each do |p|
        trabajar_proceso(p)
      end
    end
    reposicionar_cola
    guardar_buffer
  end

  private def reposicionar_cola
    posicion = 0
    @procesos.each do |p|
      if p.estado == "T"
        p.posic = 0
      else
        posicion += 1
        p.posic = posicion
      end
    end
  end

  private def procesos_terminados
    @procesos.each do |p|
      if p.estado != "T"
        return false
      end
    end
    return true
  end

  private def trabajar_proceso(p)
    if p.instru > @quantum
      p.instru -= @quantum
      p.estado = "E"
      guardar_buffer
      p.estado = p.instru == 0 ? "T" : "L"
    elsif p.estado != "T"
      p.instru = 0
      p.estado = "E"
      reposicionar_cola
      guardar_buffer
      p.estado = "T"
    end
  end

  private def guardar_archivo
    File.open(@archivo, 'a') do |f|
      f << @buffer.join
    end
    @buffer.clear
  rescue
    abort "Error al guardar #{@archivo}"
  end
end

RoundRobin.new(ARGV[0].to_i, ARGV[1].to_i, "rbtest.txt")

File: Java/RoundRobin.java
import java.io.File;
import java.io.FileWriter;
import java.io.BufferedWriter;
import java.util.ArrayList;

/**
 * Algoritmo de planificación por turnos (Round Robin)
 * @faustinoaq Sept, 2016
 */

class RoundRobin {
    int n;
    int quantum;
    String archivo;
    Proceso[] procesos;
    ArrayList<String> buffer;

    RoundRobin (int n, int quantum, String archivo) {
        this.n = n;
        this.quantum = quantum;
        this.archivo = archivo;
        this.procesos = new Proceso[n];
        this.buffer = new ArrayList<>();
        Proceso p;
        for (int i = 0; i < this.n; i++) {
            p = new Proceso();
            p.pcb = "P" + Integer.toString(i+1);
            p.id = "0" + Integer.toString(i+1);
            p.instru = 100;
            p.estado = "N";
            p.posic = i + 1;
            procesos[i] = p;
        }
        tituloBuffer();
        colaListo();
        procesarCola();
        guardarArchivo();
    }

    public static void main(String[] args) {
        new RoundRobin(Integer.parseInt(args[0]),
                       Integer.parseInt(args[1]),
                       "jatest.txt");
    }

    private void tituloBuffer() {
        this.buffer.add("Algoritmo de planificación Round Robin\n");
        this.buffer.add("N(Nuevo) L(Listo) E(Ejecución) T(Terminado)\n");
        this.buffer.add("Cantidad de procesos: " + this.n + "\n");
        this.buffer.add("Quantum: " + this.quantum + "\n");
        guardarBuffer();
    }

    private void guardarBuffer() {
        this.buffer.add("\nProceso");
        for(Proceso p : procesos) {
            this.buffer.add("\t" + p.pcb);
        }
        this.buffer.add("\nID");
        for(Proceso p : procesos) {
            this.buffer.add("\t" + p.id);
        }
        this.buffer.add("\nInstruc");
        for(Proceso p : procesos) {
            this.buffer.add("\t" + Integer.toString(p.instru));
        }
        this.buffer.add("\nEstado");
        for(Proceso p : procesos) {
            this.buffer.add("\t" + p.estado);
        }
        this.buffer.add("\nPosic");
        for(Proceso p : procesos) {
            this.buffer.add("\t" + Integer.toString(p.posic));
        }
        this.buffer.add("\n");
    }

    private void colaListo() {
        for (int i = 0; i < this.n; i++) {
            this.procesos[i].estado = "L";
        }
        guardarBuffer();
    }

    private void procesarCola() {
        while (procesosNoTerminados()) {
            for (Proceso p : procesos) {
                trabajarProceso(p);
            } 
        }
        reposicionarCola();
        guardarBuffer();
    }

    private void reposicionarCola() {
        int posicion = 0;
        for (Proceso p : procesos) {
            if (p.estado.equals("T")) {
                p.posic = 0;
            } else {
                posicion += 1;
                p.posic = posicion;
            }
        }
    }

    private boolean procesosNoTerminados() {
        for (Proceso p : procesos) {
            if (!"T".equals(p.estado)) {
                return true;
            }
        }
        return false;
    }

    private void trabajarProceso(Proceso p) {
        if (p.instru > quantum) {
            p.instru -= quantum;
            p.estado = "E";
            guardarBuffer();
            p.estado = (p.instru == 0) ? "T" : "L";
        } else if (!"T".equals(p.estado)) {
            p.instru = 0;
            p.estado = "E";
            reposicionarCola();
            guardarBuffer();
            p.estado = "T";
        }
    }

    private void guardarArchivo() {
        try {
            File f = new File(this.archivo);
            FileWriter fw = new FileWriter(f);
            BufferedWriter bw = new BufferedWriter(fw);
            bw.write(String.join("", this.buffer));
            bw.close();
            System.out.println(this.archivo + " guardado correctamente");
        } catch (Exception e) {
            System.out.println("Error al guardar " + this.archivo);
        }
    }
}
